### ISSUE TITLE
MPI likelihood computation

### ISSUE THREAD

**Issue by marziarivi** [JohannesBuchner/PyMultiNest#136]
Dear Johannes,
I would like to accelerate with MPI the computation of the likelihood because my input dataset is very large. It seems to me that the MPI version of MultiNest doesn't allow this as it just accelerate the sampler expecting the loglikelihood function to be serial. Do you confirm? My idea was to run pyMultiNest "serially" but from an MPI job where each tasks contribute to the computation of the likelihood that is finally collected and added to the task running multinest.
Basically I define a myloglikelihood function where a MPI bcast and reduce are called. However it seems it is not possible because when running the job with mpiexec, MultiNest run in parallel.
Do you think there is a way to disable this? Or do you have a better idea?
Thanks,
Marzia

**Comment by JohannesBuchner** [comment]
I think running pymultinest serially will lead to a large communication overhead. When run with MPI, N programs are run in parallel, each searching for a better fit. Multinest takes the one succeeding first, advances one iteration, and tells every process the new likelihood threshold.
In any case, I think you can stop pymultinest from using MPI by uninstalling mpi4py, but then you will (always with MPI) have N processes with their multinest running.

Maybe you can overcome the datasize issue by using shared memory, or memory maps (e.g. mmap features of hdf5 or numpy)?
Alternatively, you could try to parallelise your likelihood with thread-based techniques, for example OpenMPI, if it is written in C/C++/Fortran.

**Comment by marziarivi** [comment]
This is clear. However I need to run the job on a multi-node system, therefore using MPI, where each task reads its own dataset once and contribute to the computation of the model and likelihood at each iteration of MultiNest. Therefore the only option is, I believe, to find a way to run MultiNest serially from a master task of an MPI code. Is there no way to do that?

**Comment by JohannesBuchner** [comment]
Sounds like you want a "if rank == 0, run multinest, otherwise, go into a loop waiting for MPI input". In the rank=0 node, the likelihood function would have to send out work to the slaves.

To disable MPI of multinest, make sure libname remains 'libmultinest' (remove line 6-12 of https://github.com/JohannesBuchner/PyMultiNest/blob/master/pymultinest/run.py; also occurs when mpi4py is not installed).

**Comment by marziarivi** [comment]
Exactly. I need to run mpi4py for the main code. I made the same guess but I wasn't sure it were the only thing to do.

**Comment by JohannesBuchner** [comment]
Have a look into shmem & memmap options though if you can.
